src/client/Microsoft.Identity.Client/PlatformsCommon/Shared/InMemoryTokenCacheAccessor.cs

```diff
@@ -8,114 +8,148 @@
 using Microsoft.Identity.Client.Cache.Keys;
+using Microsoft.Identity.Client.Core;
 
 namespace Microsoft.Identity.Client.PlatformsCommon.Shared
 {
     /// <summary>
-    /// Keeps the 4 token cache dictionaries in memory. Token Cache extensions
+    /// Keeps the token cache dictionaries in memory. Token Cache extensions
     /// are responsible for persistance.
     /// </summary>
+    /// <remarks>See this post for efficient use of Concurrent Dictionary http://geekswithblogs.net/simonc/archive/2012/02/22/inside-the-concurrent-collections-concurrentdictionary.aspx</remarks>
     internal class InMemoryTokenCacheAccessor : ITokenCacheAccessor
     {
-        private readonly IDictionary<string, MsalAccessTokenCacheItem> _accessTokenCacheDictionary =
-            new Dictionary<string, MsalAccessTokenCacheItem>();
+        private readonly ConcurrentDictionary<string, MsalAccessTokenCacheItem> _accessTokenCacheDictionary =
+            new ConcurrentDictionary<string, MsalAccessTokenCacheItem>();
 
-        private readonly IDictionary<string, MsalRefreshTokenCacheItem> _refreshTokenCacheDictionary =
-            new Dictionary<string, MsalRefreshTokenCacheItem>();
+        private readonly ConcurrentDictionary<string, MsalRefreshTokenCacheItem> _refreshTokenCacheDictionary =
+            new ConcurrentDictionary<string, MsalRefreshTokenCacheItem>();
 
-        private readonly IDictionary<string, MsalIdTokenCacheItem> _idTokenCacheDictionary =
-            new Dictionary<string, MsalIdTokenCacheItem>();
+        private readonly ConcurrentDictionary<string, MsalIdTokenCacheItem> _idTokenCacheDictionary =
+            new ConcurrentDictionary<string, MsalIdTokenCacheItem>();
 
-        private readonly IDictionary<string, MsalAccountCacheItem> _accountCacheDictionary =
-            new Dictionary<string, MsalAccountCacheItem>();
+        private readonly ConcurrentDictionary<string, MsalAccountCacheItem> _accountCacheDictionary =
+            new ConcurrentDictionary<string, MsalAccountCacheItem>();
 
-        private readonly IDictionary<string, MsalAppMetadataCacheItem> _appMetadataDictionary =
-           new Dictionary<string, MsalAppMetadataCacheItem>();
+        private readonly ConcurrentDictionary<string, MsalAppMetadataCacheItem> _appMetadataDictionary =
+           new ConcurrentDictionary<string, MsalAppMetadataCacheItem>();
+        private readonly ICoreLogger _logger;
 
+        public InMemoryTokenCacheAccessor(ICoreLogger logger)
+        {
+            _logger = logger ?? throw new System.ArgumentNullException(nameof(logger));
+        }
+
+        #region Add
         public void SaveAccessToken(MsalAccessTokenCacheItem item)
         {
-            _accessTokenCacheDictionary[item.GetKey().ToString()] = item;
+            string key = item.GetKey().ToString();
+
+            // if a conflict occurs, pick the latest value
+            _accessTokenCacheDictionary.AddOrUpdate(key, item, (k, oldValue) => item);
```

Comment:
AddOrUpdate, TryRemove etc. are thread safe way of accessing a concurrent dictionary.